### DISCUSSION TITLE
2FA Token is unsafe

### DISCUSSION THREAD

**Discussion by bradbell** [community/community#75936] [Other Feature Feedback, Questions, & Ideas] [answered]
I am using FreeOTP for two factor authentication. I tried to set it up with github and got the following message:
Token is unsafe!
The token you are attempting to add contains weak cryptorgraphic parameters ... Please alert your token provider.

Should I just ignore this message an 'add anyway' ?

**Comment by ghost** [comment]
Hello @bradbell 

it's important not to brush off the warning message and just go ahead with the "add anyway" option. The message is there to alert you that the token you're using might not be secure and could be at risk of compromise. Ignoring this warning might open up the possibility of someone gaining unauthorized access to your GitHub account.

While FreeOTP is a solid choice for two-factor authentication, it's crucial to make sure you're using a robust token. A strong token typically has a lengthy secret key and employs a secure algorithm like SHA-1 or SHA-256.

If you're using FreeOTP, you can create a new token with a strong secret key by following these steps:

1. Open FreeOTP on your mobile device.
2. Tap the "+" button to set up a new token.
3. Choose "Manual entry" from the options.
4. Input a lengthy secret key. You can generate a strong one using a password generator or by typing a random string of characters.
5. Pick either "SHA-1" or "SHA-256" from the list of algorithms.
6. Tap "Add" to include the new token.

Once you've successfully generated a new token with a robust secret key, you can attempt to add it to GitHub again. At this point, the warning message about the token being unsafe should no longer pop up.

Hope this helps

**Comment by bradbell** [comment]
@giorgileladze  When I setup 2FA for pypi.org using FreeOTP, I scanned in the code provided by pypi.org and did not get a warning. When I did the same for github using FreeOTP, I got the warning, so I think that github is not providing secure tokens ?

**Comment by bradbell** [comment]
This was noticed on 2023-2-10; see
https://github.com/nextcloud/twofactor_totp/issues/1324

Is there a proper place to report this so that it gets fixed ?

**Comment by bradbell** [comment, marked as answer]
I posted a github ticket number 2441700 and they had the following reply:

"
The error message you've encountered is due to GitHub's use of an 80-bit TOTP secret which is used to ensure compatibility with older versions of Google Authenticator. FreeOTP warns about the usage of any secret that is shorter than 128-bits and is why you've encountered that error.
 
GitHub believes an 80-bit key to still be sufficient for account security, and at this time we have no plans to change this, and recommend ignoring this message if you elect to use FreeOTP for 2FA.
"

**Reply by samus-aran** [reply]
Thank you @bradbell for sharing the response for others.

**Reply by JuriJuriJuri** [reply]
thanks for sharing, had the same issue and will ignore it for now

**Comment by kworr** [comment]
What if I still feel unsure about whether adding shorter TOTP secret really improves my account security in 2024? Would it be better to provide a way to manually select better secret? This will also create a base for future rolling token upgrades.

**Reply by Matthias-Hermsdorf** [reply]
GitHub now enforces 2FA and while activating it, the Software warns about unsecurity of that process? 

I would prefer stronger security to compatibility with old versions. If you could enforce 2FA, you could send notifications to inform about a necessary update.

Or it feels like a security soap opera. If they prefer a weak security on the visible points, you don't want to take a look behind the scenes.
